topdown/http.go

```diff
@@ -568,3 +620,394 @@ func parseTimeout(timeoutVal ast.Value) (time.Duration, error) {
+	ccHeader := headers.Get("cache-control")
+
+	for _, part := range strings.Split(ccHeader, ",") {
+		part = strings.Trim(part, " ")
+		if part == "" {
+			continue
+		}
+		if strings.ContainsRune(part, '=') {
+			items := strings.Split(part, "=")
+			if len(items) != 2 {
+				continue
+			}
+			ccDirectives[strings.Trim(items[0], " ")] = strings.Trim(items[1], ",")
+		} else {
+			ccDirectives[part] = ""
+		}
+	}
+
+	return ccDirectives
+}
+
+func getResponseHeaderDate(headers http.Header) (date time.Time, err error) {
+	dateHeader := headers.Get("date")
+	if dateHeader == "" {
+		err = fmt.Errorf("no date header")
+		return
+	}
+	return time.Parse(time.RFC1123, dateHeader)
+}
+
+func getResponseHeaderExpires(headers http.Header) (date time.Time, err error) {
+	expiresHeader := headers.Get("expires")
+	if expiresHeader == "" {
+		return
+	}
+	return time.Parse(time.RFC1123, expiresHeader)
+}
+
+// parseMaxAgeCacheDirective parses the max-age directive expressed in delta-seconds as per
+// https://tools.ietf.org/html/rfc7234#section-1.2.1
+func parseMaxAgeCacheDirective(cc map[string]string) (deltaSeconds, error) {
+	maxAge, ok := cc["max-age"]
+	if !ok {
+		return deltaSeconds(-1), nil
+	}
+
+	val, err := strconv.ParseUint(maxAge, 10, 32)
+	if err != nil {
+		if numError, ok := err.(*strconv.NumError); ok {
+			if numError.Err == strconv.ErrRange {
+				return deltaSeconds(math.MaxInt32), nil
+			}
+		}
+		return deltaSeconds(-1), err
+	}
+
+	if val > math.MaxInt32 {
+		return deltaSeconds(math.MaxInt32), nil
+	}
+	return deltaSeconds(val), nil
+}
+
+func formatHTTPResponseToAST(resp *http.Response, forceJSONDecode bool) (ast.Value, int, error) {
+
+	var resultBody interface{}
+	var resultRawBody []byte
+
+	var buf bytes.Buffer
+	tee := io.TeeReader(resp.Body, &buf)
+	resultRawBody, err := ioutil.ReadAll(tee)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// If the response body cannot be JSON decoded,
+	// an error will not be returned. Instead the "body" field
+	// in the result will be null.
+	if isContentTypeJSON(resp.Header) || forceJSONDecode {
+		json.NewDecoder(&buf).Decode(&resultBody)
+	}
+
+	result := make(map[string]interface{})
+	result["status"] = resp.Status
+	result["status_code"] = resp.StatusCode
+	result["body"] = resultBody
+	result["raw_body"] = string(resultRawBody)
+	result["headers"] = getResponseHeaders(resp.Header)
+
+	resultObj, err := ast.InterfaceToValue(result)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return resultObj, len(resultRawBody), nil
+}
+
+func getResponseHeaders(headers http.Header) map[string]interface{} {
+	respHeaders := map[string]interface{}{}
+	for headerName, values := range headers {
+		var respValues []interface{}
+		for _, v := range values {
+			respValues = append(respValues, v)
+		}
+		respHeaders[strings.ToLower(headerName)] = respValues
+	}
+	return respHeaders
+}
+
+// HTTPRequestExecutor defines an interface for the http send cache
+type HTTPRequestExecutor interface {
```

Comment:
Let's not export this interface from the package because it's a detail of the built-in function implementation.

Reply:
Done